From 6334154184e7598a0264f0d70cc27994ef3feb93 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 18 Sep 2020 15:43:57 +0800
Subject: [PATCH 40/81] dri2: Fix sw cursor rendering issue

When reporting damage, the misprite swcursor will try to remove cursor
by restoring the under cursor area.

This will override the previous draw, so we need to move it before
the rendering.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 hw/xfree86/drivers/modesetting/dri2.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/dri2.c b/hw/xfree86/drivers/modesetting/dri2.c
index 79bbd60a3..4a5ed28bd 100644
--- a/hw/xfree86/drivers/modesetting/dri2.c
+++ b/hw/xfree86/drivers/modesetting/dri2.c
@@ -622,8 +622,6 @@ ms_dri2_exchange_buffers(DrawablePtr draw, DRI2BufferPtr front,
     *front_pix = *back_pix;
     *back_pix = tmp_pix;
 
-    ms_exchange_buffers(front_priv->pixmap, back_priv->pixmap);
-
     /* Post damage on the front buffer so that listeners, such
      * as DisplayLink know take a copy and shove it over the USB.
      */
@@ -632,6 +630,9 @@ ms_dri2_exchange_buffers(DrawablePtr draw, DRI2BufferPtr front,
     region.extents.y2 = front_priv->pixmap->drawable.height;
     region.data = NULL;
     DamageRegionAppend(&front_priv->pixmap->drawable, &region);
+
+    ms_exchange_buffers(front_priv->pixmap, back_priv->pixmap);
+
     DamageRegionProcessPending(&front_priv->pixmap->drawable);
 }
 
-- 
2.38.5

